### PR TITLE
CR-1125007 hwmon_sdm: add changes to report sensor status

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Added changes to report the sensor status as part of hwmon sysfs nodes named "*_status".
Followed doc https://confluence.xilinx.com/display/DCG/Versal+In-band+sensor+data+reporting for reporting the sensor status as per their value.

Signed-off-by: Rajkumar Rampelli <rajkumar@xilinx.com>

#### Problem solved by the commit
VMR sends sensor status for each sensor. hwmon_sdm driver should read this value and store it as part of hwmon sysfs nodes.
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
CR-1125007
#### How problem was solved, alternative solutions (if any) and why they were rejected
Added changes in hwmon_sdm driver and relevant shim layer of XRT
#### Risks (if any) associated the changes in the commit
NA
#### What has been tested and how, request additional testing if necessary
hwmon sysfs access of these newly added sensor status
#### Documentation impact (if any)
NA

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
-bash-4.2# ls /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/
curr1_average  curr2_max      curr4_input   in0_max      in2_input    in3_status   in5_label       power1_max     temp2_average  temp3_max      temp5_input
curr1_input    curr2_status   curr4_label   in0_status   in2_label    in4_average  in5_max         power1_status  temp2_input    temp3_status   temp5_label
curr1_label    curr3_average  curr4_max     in1_average  in2_max      in4_input    in5_status      subsystem      temp2_label    temp4_average  temp5_max
curr1_max      curr3_input    curr4_status  in1_input    in2_status   in4_label    name            temp1_average  temp2_max      temp4_input    temp5_status
curr1_status   curr3_label    device        in1_label    in3_average  in4_max      power           temp1_input    temp2_status   temp4_label    uevent
curr2_average  curr3_max      in0_average   in1_max      in3_input    in4_status   power1_average  temp1_label    temp3_average  temp4_max
curr2_input    curr3_status   in0_input     in1_status   in3_label    in5_average  power1_input    temp1_max      temp3_input    temp4_status
curr2_label    curr4_average  in0_label     in2_average  in3_max      in5_input    power1_label    temp1_status   temp3_label    temp5_average

-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/in1_status                                                                                         Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/in2_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/in3_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/in4_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/in5_status
Sensor Present and Valid

-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/curr1_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/curr2_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/curr3_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/curr4_status
Sensor Present and Valid

-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/power1_status
Sensor Present and Valid

-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/in0_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/temp1_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/temp2_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/temp3_status
Sensor Present and Valid
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/temp4_status
Sensor Not Present
-bash-4.2# cat /sys/bus/pci/devices/0000:86:00.0/hwmon/hwmon3/temp5_status
Sensor Not Present

**Review by stsoe — CHANGES_REQUESTED** [review]
Please re-factor get_sdm_sensors so that no function is more than 5-10 lines of code.  The current function obviously doesn't scale.

while -> for -> if  -> if, is just too much.   Completely unreadable.  If the code is refactored and functions named appropriately, the code will be self-documenting.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c [hunk 1]
```diff
@@ -318,6 +319,25 @@ static ssize_t hwmon_sensor_show(struct device *dev,
 		}
 		memcpy(&uval, &sdm->sensor_data[repo_id][buf_index], buf_len);
 		sz = sprintf(buf, "%u\n", uval);
+	} else if (field_id == SYSFS_SDR_STATUS_VAL) {
+		memcpy(&uval, &sdm->sensor_data[repo_id][buf_index], buf_len);
+		switch(uval) {
+		case 0x00:
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c:325]: Magic numbers.  Can these values be defined earlier.  For example:
```
enum {
  SD_NOT_PRESENT = 0,
  SD_PRESENT = 0x01,
  SD_DATA_NOT_AVAILABLE = 0x02,
  SD_DEFAULT_VALUE = 0x7F
} SensorData;
```

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c [hunk 2]
```diff
@@ -687,6 +713,14 @@ static int parse_sdr_info(char *in_buf, struct xocl_hwmon_sdm *sdm, bool create_
 						xocl_err(&sdm->pdev->dev, "Unable to create sysfs node (%s), err: %d\n", sysfs_name[3], err);
 					}
 				}
+
+				//Create *_status sysfs node
+				if(strlen(sysfs_name[4]) != 0) {
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c:718]: Magic number.  What is so special with the 4th index element.

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 3]
```diff
@@ -139,17 +139,72 @@ struct sdm_sensor_info
...
+  static data_type
+  parse_sysfs_nodes(const xrt_core::device* device,
+                    std::string tpath,
+                    bool *next_id)
+  {
+    auto pdev = get_pcidev(device);
+    //sensors are stored in hwmon sysfs dir with name ends with as follows.
+    std::array<std::string, 6> sname_end = {"label", "input", "max", "average", "highest", "status"};
+    int max_end_types = sname_end.size();
+    data_type data;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:151]: Unitialized variables.

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 4]
```diff
@@ -139,17 +139,72 @@ struct sdm_sensor_info
...
+
+    for (int end_id = 0; end_id < max_end_types; end_id++)
```
Comment (rozumx) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:153]: This code can be simplified with early "continue" exits.  For example:
```
for (int end_id = 0; end_id < max_end_types; end_id++)
    {
      std::string target_snode = tpath + sname_end[end_id];
      if (end_id == 0) {
        std::string errmsg;
        std::string label;
        pdev->sysfs_get("", target_snode, errmsg, label);
        if (!errmsg.empty())  {
          //go and read next sysfs node
          *next_id = true;
          end_id = max_end_types;
        } else {
          data.label = label;
       }
       continue;
      }

     if (end_id == 5)     {
        std::string errmsg;
        std::string status;
        pdev->sysfs_get("", target_snode, errmsg, status);
        if (errmsg.empty())
           data.status = status;
        continue;
      }

      std::string errmsg;
      uint32_t input = 0;
      pdev->sysfs_get<uint32_t>("", target_snode, errmsg, input, EINVAL);
      if (!errmsg.empty())
        continue;

      if (end_id == 1)
        data.input = input;
      else if (end_id == 2)
        data.max = input;
      else if (end_id == 3)
        data.average = input;
      else
        data.highest = input;
    }
```

**Review by rozumx — CHANGES_REQUESTED** [review]

**Review by rozumx — APPROVED** [review]
Looks good. Thank you for making the changes.

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by rajkumar-xilinx** [comment]
Hi @maxzhen and @stsoe , could you please review this change ? thank you

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 5]
```diff
@@ -165,51 +221,15 @@ struct sdm_sensor_info
...
-          else if (end_id == 3)
-            data.average = input;
-          else
-            data.highest = input;
-        }
-      } // for (end_id =0; end_id < max ...)
-
-      if (data.label.compare("N/A"))
-        output.push_back(data);
+      std::string tpath = path + slash + type + std::to_string(start_id) + underscore;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:230]: unreadable, any reason `slash` and `underscore` are symbolic?  Any `type` incurs a string copy, why?

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 6]
```diff
@@ -139,17 +139,73 @@ struct sdm_sensor_info
...
+    auto pdev = get_pcidev(device);
+    //sensors are stored in hwmon sysfs dir with name ends with as follows.
+    std::array<std::string, 6> sname_end = {"label", "input", "max", "average", "highest", "status"};
+    int max_end_types = sname_end.size();
+    data_type data = { 0 };
+
+    for (int end_id = 0; end_id < max_end_types; end_id++)
+    {
+      std::string target_snode = tpath + sname_end[end_id];
+      if (end_id == 0)
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:156]: // label case I guess?
The numbers really obfuscate and prevent this code from being readable.  Comments would help.

**Review by stsoe — CHANGES_REQUESTED** [review]

**Comment by rajkumar-xilinx** [comment]
Hi @stsoe , Could you please review this PR again and let me know, thank you.

**Review by stsoe — APPROVED** [review]
I don't think parse_sysfs_nodes scales.  It is more or less unreadable, but sufficiently short to not be a blocking issue.  Approved.

**Pull request merged**